### ISSUE TITLE
V51 OAuth: Consider narrowing or expanding the scope for the OAuth2 chapter

### ISSUE THREAD

**Issue by TobiasAhnoff** [OWASP/ASVS#1924]

**Comment by elarlang** [comment]
From pentesting experience, everyone using OAuth/OIDC and only a few know, what those are, why and for what purpose they exist, and how they should be used.

To find the balance between details and abstractation is hard job. The situation at the moment is, that the entire chapter - including chapter texts and requirements - is open to feedback and recommendation.

Personally, I would like to see OAuth and OIDC covered and clearly separated to point out, that Authorization and authentication tools work hand-in-hand, but those are not the same tools and one should not mix them. A widespread problem is that OAuth is used for authentication, as I pointed out in another issue https://github.com/OWASP/ASVS/issues/1916 "Discussion/Proposal 4".

For the summary - if you have a vision, of how it should be organized, please share it! :)

ping @tghosth @jmanico @csfreak92

**Comment by jmanico** [comment]
> Personally, I would like to see OAuth and OIDC covered and clearly separated...

Agreed, Elar.

**Comment by tghosth** [comment]
I agree that it would be good to cover both OAuth2 and OIDC.

There will be a tricky balance between too much detail and too little detail. I think we can refer to supporting documentation if necessary. For example in 5.0 the logging chapter is likely to refer to the OWASP logging cheatsheet and the password storage section is likely to refer to the password storage cheatsheet.

@TobiasAhnoff  @csfreak92 what are your thoughts?

**Comment by csfreak92** [comment]
I agree it would be good to cover both OAuth2 (and potentially working on OAuth2.1) and OIDC, but we need to find the right balance between too much detail and too little detail, which is something we try to deal with. :) 

Maybe we can loop in @jsherm-fwdsec as he had initial ideas for OIDC in one of our issues particularly #1826. Any thoughts Jordan?

**Comment by elarlang** [comment]
The ping-pong here in the issues is at a really slow pace. It is possible also to organize some video calls and have quick discussions and overviews, of what is done, what should be done, what the expectations and abstract goals, etc.

**Comment by TobiasAhnoff** [comment]
Sorry for the late reply, I started working on it, but other things got in the way....I would be happy to attend a video call meeting and I plan to reply here during this week...

**Comment by jsherm-fwdsec** [comment]
Thanks @csfreak92. I'd agree that OAuth/OIDC need to be split out in to their own respective sections. @elarlang happy to attend a video call as well if that would help figure out the scope and perhaps we can divide and conquer.

**Comment by elarlang** [comment]
I try to give answers to the questions presented in the comment.

> What do we want to achieve by adding one OAuth2 and one OIDC chapter to ASVS?

At the moment we can say, that it is not "hardcoded" or clearly defined, the goal is to find it out.

If we take a look at OWASP Top 10 2021, A01 is Broken Access Control, and A05 is security misconfiguration. I have not investigated the stats behind that, but I assume that misconfigured and incorrect usage of OAuth is an important "donor" for those positions.


> Then ASVS must decide on which flows to include in the scope and make that clear from the start of the OAuth2 and OIDC chapters. Should ASVS only focus on classic browser based apps or should mobile clients be include? Should "detached" flows (device grant and CIBA for OIDC) be included?

Although nowadays browser-based flows are widespread for the 1st party usage, OAuth was not built for this kind of usage. I'm not convinced that is the correct way to use or to allow to use OAuth. Often it is unnecessary overengineering and creating many vulnerabilities due to misconfigurations.

For ASVS, we are not limited only to browser-to-server communication, service-to-service must be covered as well - the actual goal for the OAuth.

> .. should OIDC id-token flow be included?

and

> Perhaps the OIDC id-token flow should be included from the start, since it is commonly used and address the login scenario where you don´t want access tokens (and involve OAuth).

Yes, often OAuth is used for authentication and OIDC should be used instead.


> The client-credentials flow should perhaps be included in other ASVS chapters for "service integrations" and authorization or should it be part of this chapter?

I think OAuth specific requirements should all be in one place.


> The second goal is the hard one. For OIDC, FAPI kind of captures what ASVS could look like, so it might be hard to avoid being some sort of copy of that...but we don´t know until we tried!

By principle, it is not a problem to "copy" the content - there are arguments and research behind that. Even if you do your own, you should probably reach the same result and the end result should be really similar. In the same way, many requirements in the authentication and session management come from NIST.

Another question - should we duplicate that, or should we have abstract requirement "Verify that OIDC setup and configuration is ...".


> For OAuth2, I don´t think there is such short and percise guidance as FAPI, so in a way ASVS could fill that gap, But it is challenging with the level of detaial and how many specs that should be in scope, should e g RAR be included?

My idea is to address frequently happening and common mistakes with separate requirements. We are not able to cover all the OAuth extra-extra-n-extra level security mechanisms. In reality, there are so many separate add-on specifications in development, that in a year of ASVS release (whenever it will happen), it will be outdated already.

> One mistake to make when implementing OAuth2/OIDC solutions is to build things on your own...In my opinion, I think it would be good if ASVS provides guidance on topics like this, based on experiences, it is not obvious when reading the specs that building your own or not protecting the signing-key (e g by sharing it with resource servers) introduce risk.

ASVS may address widespread problems with separate requirement, but the goal and the scope for the ASVS is not to provide guidance.

**Comment by elarlang** [comment]
ping @csfreak92 @TobiasAhnoff @jsherm-fwdsec 

We should find a common communication channel to discuss suitable times for video calls or just chat on the topic. I think the best option is to use the OWASP Slack for that. Please find your way there (if you are not there yet) and I'll open a separate channel for that.

**Comment by jsherm-fwdsec** [comment]
> ping @csfreak92 @TobiasAhnoff @jsherm-fwdsec
> 
> We should find a common communication channel to discuss suitable times for video calls or just chat on the topic. I think the best option is to use the OWASP Slack for that. Please find your way there (if you are not there yet) and I'll open a separate channel for that.

I started a thread on OWASP Slack :)

**Comment by TobiasAhnoff** [comment]
> > ping @csfreak92 @TobiasAhnoff @jsherm-fwdsec
> > We should find a common communication channel to discuss suitable times for video calls or just chat on the topic. I think the best option is to use the OWASP Slack for that. Please find your way there (if you are not there yet) and I'll open a separate channel for that.
> 
> I started a thread on OWASP Slack :)

Slack will be good! But I´m not in the OWASP Slack, can you send an invite?

**Comment by TobiasAhnoff** [comment]
> > ping @csfreak92 @TobiasAhnoff @jsherm-fwdsec
> > We should find a common communication channel to discuss suitable times for video calls or just chat on the topic. I think the best option is to use the OWASP Slack for that. Please find your way there (if you are not there yet) and I'll open a separate channel for that.
> 
> I started a thread on OWASP Slack :)

I have joined the OWASP Slack, but I cant find the thread :) 
ping @jsherm-fwdsec

**Comment by elarlang** [comment]
@TobiasAhnoff - for me it seems the issue can be closed, as risen topics are on the move and here is no further action to take.

**Comment by TobiasAhnoff** [comment]
Yes, this has been covered in other issues.
